datafusion/functions/src/core/named_struct.rs

```diff
@@ -0,0 +1,196 @@
+            };
+
+            Ok((name, chunk[1].clone()))
+        })
+        .collect::<Result<Vec<_>>>()?
+        .into_iter()
+        .unzip();
+
+    let arrays = ColumnarValue::values_to_arrays(&values)?;
+
+    let fields = names.into_iter()
+        .zip(arrays)
+        .map(|(name, value)| {
+            (
+                Arc::new(Field::new(
+                    name,
+                    value.data_type().clone(),
+                    true,
+                )),
+                value,
+            )
+        })
+        .collect::<Vec<_>>();
+
+    Ok(ColumnarValue::Array(Arc::new(StructArray::from(fields))))
+}
+
+#[derive(Debug)]
+pub(super) struct NamedStructFunc {
+    signature: Signature,
+}
+
+impl NamedStructFunc {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::variadic_any(Volatility::Immutable),
+        }
+    }
+}
+
+impl ScalarUDFImpl for NamedStructFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "named_struct"
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, _arg_types: &[DataType]) -> Result<DataType> {
+        internal_err!("named_struct: return_type called instead of return_type_from_exprs")
+    }
+
+    fn return_type_from_exprs(
+        &self,
+        args: &[datafusion_expr::Expr],
+        schema: &dyn datafusion_common::ExprSchema,
+        _arg_types: &[DataType],
+    ) -> Result<DataType> {
+        // do not accept 0 arguments.
+        if args.is_empty() {
+            return exec_err!("named_struct requires at least one pair of arguments, got 0 instead");
+        }
+
+        if args.len() % 2 != 0 {
+            return exec_err!("named_struct requires an even number of arguments, got {} instead", args.len());
+        }
+
+        let return_fields = args
+            .chunks_exact(2)
+            .enumerate()
+            .map(|(i, chunk)| {
+                let name = &chunk[0];
+                let value = &chunk[1];
+
+                if let Expr::Literal(ScalarValue::Utf8(Some(name))) = name {
+                    Ok(Field::new(name, value.get_type(schema)?, true))
+                } else {
+                    exec_err!("named_struct even arguments must be string literals, got {name} instead at position {}", i * 2)
+                }
+            })
+            .collect::<Result<Vec<Field>>>()?;
+        Ok(DataType::Struct(Fields::from(return_fields)))
+    }
+
+    fn invoke(&self, args: &[ColumnarValue]) -> Result<ColumnarValue> {
+        named_struct_expr(args)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use arrow::array::Int64Array;
+    use datafusion_common::cast::as_struct_array;
+    use datafusion_common::ScalarValue;
+
+    #[test]
+    fn test_named_struct() {
+        // named_struct("first", 1, "second", 2, "third", 3) = {"first": 1, "second": 2, "third": 3}
```

Comment:
I think we can have this test in `struct.slt` too

Reply:
I agree -- let's do it as a follow on PR

Reply:
I was pushing commits to this PR anyways as the CI failed, so I also did this suggestion in f35cddc12

Reply:
Thanks @alamb 🙏